### PR TITLE
compiler: Fix some vestigial +0.0/-0.0 issues

### LINKED ISSUE #7901: [erlc] Consistency check failure related to -0.0/+0.0: "bad_arg_type"
**Describe the bug**
Running `erlc` on the following code:
```
-export([start/0]).

f1()  -> 
	f6(catch (true = ok), -0.0).

f3()  -> 
	f6(ok, 0.0).

f6(_, _)  -> 
	ok.

start()  -> 
	f1(),
	f3(),
	f6(ok, ok).
```
Results in the following error:
```
test745_mini:1: function f1/0+12:
  Internal consistency check failed - please report this bug.
  Instruction: {call_last,2,{f,8},1}
  Error:       {bad_arg_type,{x,1},
                             {t_float,{-0.0,-0.0}},
                             {t_union,{t_atom,[ok]},
                                      none,
                                      {t_float,{0.0,0.0}},
                                      none,none}}:
```

**Affected versions**
master

**Additional context**
It looks possibly related to #7168 and the OTP 27 change that will make `+0.0 =:= -0.0` return false.

### PR DESCRIPTION
**Pull request by jhogberg**
Fixes #7901

### CODE REVIEW

File: lib/compiler/src/beam_types.erl [hunk 1]
```diff
@@ -1218,21 +1221,104 @@ number_from_range(N) ->
...
+    '-inf';
+inf_min(A, B) when is_float(A), is_float(B) ->
+    case float_comparable(A) =< float_comparable(B) of
+        true -> A;
+        false -> B
+    end;
+inf_min(A, B) when A =< B ->
+    A;
+inf_min(A, B) when A > B ->
+    B.
```
Comment (bjorng) [inline lib/compiler/src/beam_types.erl:1252]:
```suggestion
inf_min(A, B) ->
    min(A, B).
```

File: lib/compiler/src/beam_types.erl [hunk 2]
```diff
@@ -1218,21 +1221,104 @@ number_from_range(N) ->
...
+    A;
+inf_max(A, B) when is_float(A), is_float(B) ->
+    case float_comparable(A) >= float_comparable(B) of
+        true -> A;
+        false -> B
+    end;
+inf_max(A, B) when A >= B ->
+    A;
+inf_max(A, B) when A < B ->
+    B.
```
Comment (bjorng) [inline lib/compiler/src/beam_types.erl:1266]:
```suggestion
inf_max(A, B) ->
    max(A, B).
```

**Review by bjorng — APPROVED** [review]

**Comment by RobinMorisset** [comment]
Thanks for this quick fix.
Some variants of the original bug remain, e.g.:
```
-export([f2/0, start/0]).

f2() ->
    [ok || _ := _ <- ok, f3(catch f3(ok, -1), -0.0)].

f3(_, 18446744073709551615) ->
    ok.

start() ->
    f3(ok, ok).
```
Still crashes with
```
test98296:1: function '-f2/0-lc$^0/1-0-'/1+17:
  Internal consistency check failed - please report this bug.
  Instruction: {call_last,2,{f,9},1}
  Error:       {bad_arg_type,{x,1},
                             {t_float,{-0.0,-0.0}},
                             {t_union,{t_atom,[ok]},
                                      none,
                                      {t_number,{-1,0}},
                                      none,none}}:
```

**Comment by jhogberg** [comment]
Thanks, I'll get right on it :)

**Comment by RobinMorisset** [comment]
This PR also seems to have introduced a new bug:
```
-export([f4/1]).

f4(_V1) ->
    -0.0 = abs(_V1).
```
Crashes erlc with
```
Function: f4/1
Sub pass ssa_opt_type_start
/home/rmorisset/minimized/test98057.erl: internal error in pass beam_ssa_opt:
exception error: no match of right hand side value false
  in function  beam_types:glb_ranges/2 (beam_types.erl, line 933)
  in call from beam_types:glb/2 (beam_types.erl, line 879)
  in call from beam_ssa_type:meet_types/2 (beam_ssa_type.erl, line 2896)
  in call from beam_ssa_type:infer_types_br_1/3 (beam_ssa_type.erl, line 2512)
  in call from beam_ssa_type:infer_types_br/4 (beam_ssa_type.erl, line 2477)
  in call from beam_ssa_type:update_successors/5 (beam_ssa_type.erl, line 1968)
  in call from beam_ssa_type:sig_bs/8 (beam_ssa_type.erl, line 258)
  in call from beam_ssa_type:sig_function_1/4 (beam_ssa_type.erl, line 221)
```
but only with this PR.

**Comment by RobinMorisset** [comment]
> This PR also seems to have introduced a new bug:

Two other testcases with the same symptom:
```
-export([f7/1]).

f7(_V1) when -0.0 < floor(_V1 band 1) ->
    ok.
```
and
```
-export([f9/0]).

f9() ->
    -0.0 =
        case
            fun() ->
                ok
            end
        of
            _V2 when (_V2 > ok) ->
                2147483647.0;
            _ ->
                -2147483648
        end * 0.
```

**Comment by jhogberg** [comment]
Thanks, I had introduced some inconsistencies in the lattice, I've pushed up a new version now :)

**Comment by RobinMorisset** [comment]
erlfuzz has run for more than 1h with no bug found (it was finding one every 5 to 10mn before this latest fix), so it seems to be fixed for good, thanks!

**Pull request merged**